### DISCUSSION TITLE
In PP or VPP, display matrix results with changeable size

### DISCUSSION THREAD

**Discussion by sSajjad90** [idaholab/moose#24161] [Q&A General] [answered]
Hello,
I'm attempting to develop a material code '*.C' file to determine the number density distribution of three different phases. For each iteration, the results are vectors like [a_11, a_12,...,a_1n] for each phase ( a 3*n matrix for three phases), and I want to return them to the app as follows: 

Declaration of material property in *.C file:
` a1_calc(declareProperty<std::vector<std::vector<Real>>>("a1_calc")),`

The  relevant part of resizing the vector and assigning values:

```
 a1_calc[_qp].resize(Phase_Num);
  for (unsigned int i = 0; i < Phase_Num; i++)
  {
    a1_calc[_qp][i].resize((*(InitAtomNum[i]))[_qp]);
    for (size_t j = 0; j < ((*(InitAtomNum[i]))[_qp]); j++) {
      a1_calc[_qp][i][j] = 0;
    }
  for (unsigned int i = 0; i < Phase_Num; i++)
  {
    for (size_t j = 0; j < ((*(InitAtomNum[i]))[_qp]); j++) {
      a1_calc[_qp][i][j] = InitNumDensitiesVector[i][j];
     }
  }
```

I'd want to have a matrix of results (3*n) returned by 'a1_calc' in postprocessor, or three postprocessor vectors for each phase, comprising all vector values in each iteration.
Please keep in mind that the size of vectors varies with each programme iteration. (For example, in the first iteration,it would be 12, and subsequently shift to 30 in the third iteration or 3000 in the 100th iteration.
 I was wondering how I might present this type of result in VPP.

Thanks in advance for your kind support,

Sincerely,
Sajjad

**Comment by GiudGiud** [comment]
Hello

VPP aren’t meant to deal with vectors of vectors. 
you ll either want to work on a Reporter, which is a more general but more complex object than a VPP 

or simply define multiple VPPs for each vector within the vector of vectors 

Guillaume

**Reply by sSajjad90** [reply]
Thank you very much.
That's precisely what I've been stuck on.  If I define three independent vectors instead of a vector of vectors. I'm still not sure what 'AuxVariables,' 'AuxKernels,' and 'VPP (or Reporter)' might be used to read all dates from a vector with varying sizes.

For instance, if there is just a defined vector, supply a vector of values in each iteration, as seen below. 
Would you please kindly tell me which 'AuxVariables,' 'AuxKernels,' and 'VPP' are capable of aggregating and displaying all values for each iteration?

```
[Materials]
  [./d]
  .
  .
  .
  a1_calc = 'a1_calc'
  [../]
[ ]
Results:
t= 1: a1_calc = {1, 2, 3, 4}
t= 2: a1_calc = {6, 14, 3, 4, 22, 13, 18}
t= 3: a1_calc = {56, 14, 3, 4, 72, 13, 48, 16, 0, 0, 12, 13}
t= n: a1_calc = {a_1, ...a_m}
```

**Reply by GiudGiud** [reply, marked as answer]
'AuxVariables,' are meant for data sized with the number of elements/nodes in the mesh. So this is not what you want if i understand correctly. 

A VPP should be able to do this. 
(and now that I've taken a closer look, multiple vectors are supported by VPPs! As long as the number of vectors is known at construction as far as I can tell)
Some examples of dynamically sized vector postprocessors are:

- the samples, derived from SamplerBase. The size is directed by the derived class, the number of nodes for the NodalValueSampler for example
- the eigenvalues VPP

A lot of the other VPPs have their vectors sized on initialization. It would not be challenging to size on execute instead.
